File: aihound/scanners/vscode_extensions.py
"""Generic scanner for secrets in VS Code extension globalStorage directories."""

from __future__ import annotations

import json
import logging
from pathlib import Path
from typing import Iterator

from aihound.core.scanner import (
    BaseScanner, CredentialFinding, ScanResult, StorageType,
)
from aihound.core.platform import (
    detect_platform, Platform, get_home, get_appdata, get_wsl_windows_home,
    get_xdg_config,
)
from aihound.core.redactor import mask_value
from aihound.core.permissions import (
    get_file_permissions, get_file_owner, assess_risk,
    get_file_mtime, describe_staleness,
)
from aihound.remediation import hint_manual
from aihound.scanners import register

logger = logging.getLogger("aihound.scanners.vscode_extensions")

# Extensions already covered by dedicated scanners — skip to avoid double reporting.
EXCLUDED_EXTENSIONS = {
    "github.copilot",
    "github.copilot-chat",
    "saoudrizwan.claude-dev",  # Cline
}

# Maximum JSON file size we're willing to parse (avoid data blobs, indexes, etc.)
MAX_JSON_SIZE = 1024 * 1024  # 1 MB

# Max recursion depth when walking extension directories / JSON structures.
MAX_DIR_DEPTH = 3
MAX_JSON_DEPTH = 8

SECRET_KEY_SUBSTRINGS = (
    "token", "key", "secret", "password", "apikey", "auth",
)

REMEDIATION = (
    "Use VS Code's SecretStorage API or OS keychain for extension credentials"
)


@register
class VSCodeExtensionsScanner(BaseScanner):
    def name(self) -> str:
        return "VS Code Extensions"

    def slug(self) -> str:
        return "vscode-extensions"

    def scan(self, show_secrets: bool = False) -> ScanResult:
        plat = detect_platform()
        result = ScanResult(scanner_name=self.name(), platform=plat.value)

        for base in self._get_global_storage_roots(plat):
            self._scan_global_storage(base, result, show_secrets)

        return result

    def _get_global_storage_roots(self, plat: Platform) -> list[Path]:
        roots: list[Path] = []

        if plat in (Platform.LINUX, Platform.WSL):
            roots.append(get_xdg_config() / "Code" / "User" / "globalStorage")
        if plat == Platform.MACOS:
            roots.append(
                get_home() / "Library" / "Application Support" / "Code"
                / "User" / "globalStorage"
            )
        if plat == Platform.WINDOWS:
            appdata = get_appdata()
            if appdata:
                roots.append(appdata / "Code" / "User" / "globalStorage")

        if plat == Platform.WSL:
            # Also scan Windows-side VS Code globalStorage via /mnt/c
            appdata = get_appdata()
            if appdata:
                roots.append(appdata / "Code" / "User" / "globalStorage")
            win_home = get_wsl_windows_home()
            if win_home:
                roots.append(
                    win_home / "AppData" / "Roaming" / "Code" / "User" / "globalStorage"
                )

        # De-duplicate while preserving order
        seen: set[str] = set()
        unique: list[Path] = []
        for r in roots:
            key = str(r)
            if key not in seen:
                seen.add(key)
                unique.append(r)
        return unique

    def _scan_global_storage(
        self, base: Path, result: ScanResult, show_secrets: bool
    ) -> None:
        if not base.exists():
            logger.debug("globalStorage not found: %s", base)
            return

        logger.debug("Scanning globalStorage root: %s", base)
        try:
            entries = list(base.iterdir())
        except OSError as e:
            logger.debug("Could not list %s: %s", base, e)
            return

        for entry in entries:
            if not entry.is_dir():
                continue
            extension_id = entry.name
            if extension_id in EXCLUDED_EXTENSIONS:
                continue

            for json_file in self._iter_json_files(entry, depth=0):
                self._scan_extension_json(
                    json_file, extension_id, result, show_secrets,
                )

    def _iter_json_files(self, directory: Path, depth: int) -> Iterator[Path]:
        if depth > MAX_DIR_DEPTH:
            return
        try:
            children = list(directory.iterdir())
        except OSError:
            return

        for child in children:
            try:
                if child.is_file() and child.suffix.lower() == ".json":
                    try:
                        if child.stat().st_size <= MAX_JSON_SIZE:
                            yield child
                    except OSError:
                        continue
                elif child.is_dir():
                    yield from self._iter_json_files(child, depth + 1)
            except OSError:
                continue

    def _scan_extension_json(
        self, path: Path, extension_id: str,
        result: ScanResult, show_secrets: bool,
    ) -> None:
        logger.debug("Reading extension JSON: %s", path)
        perms = get_file_permissions(path)
        owner = get_file_owner(path)
        mtime = get_file_mtime(path)

        try:
            data = json.loads(path.read_text(encoding="utf-8"))
        except (json.JSONDecodeError, OSError) as e:
            logger.debug("Could not parse %s: %s", path, e)
            return

        staleness_note = f"File last modified: {describe_staleness(mtime)}" if mtime else None

        for key_path, value in self._walk(data, prefix="", depth=0):
            # Only look at string values that match a secret-looking key
            if not isinstance(value, str):
                continue

            leaf_key = key_path.rsplit(".", 1)[-1].lower()
            if not any(sub in leaf_key for sub in SECRET_KEY_SUBSTRINGS):
                continue

            if not self._value_looks_secret(value):
                continue

            notes = [f"Extension: {extension_id}", f"JSON path: {key_path}"]
            if staleness_note:
                notes.append(staleness_note)

            storage = StorageType.PLAINTEXT_JSON
            result.findings.append(CredentialFinding(
                tool_name=self.name(),
                credential_type=f"{extension_id}:{key_path}",
                storage_type=storage,
                location=str(path),
                exists=True,
                risk_level=assess_risk(storage, path),
                value_preview=mask_value(value, show_full=show_secrets),
                raw_value=value if show_secrets else None,
                file_permissions=perms,
                file_owner=owner,
                file_modified=mtime,
                remediation=REMEDIATION,
                remediation_hint=hint_manual(REMEDIATION),
                notes=notes,
            ))

    def _walk(self, data, prefix: str, depth: int):
        """Yield (key_path, value) for every scalar in the JSON tree."""
        if depth > MAX_JSON_DEPTH:
            return
        if isinstance(data, dict):
            for k, v in data.items():
                new_prefix = f"{prefix}.{k}" if prefix else str(k)
                if isinstance(v, (dict, list)):
                    yield from self._walk(v, new_prefix, depth + 1)
                else:
                    yield new_prefix, v
        elif isinstance(data, list):
            for i, v in enumerate(data):
                new_prefix = f"{prefix}[{i}]" if prefix else f"[{i}]"
                if isinstance(v, (dict, list)):
                    yield from self._walk(v, new_prefix, depth + 1)
                else:
                    yield new_prefix, v

    @staticmethod
    def _value_looks_secret(value: str) -> bool:
        """Length 20+, mostly alphanumeric, not a path or URL."""
        if len(value) < 20:
            return False
        if value.startswith(("/", "\\", "http://", "https://", "file://")):
            return False
        # Reject things that clearly aren't tokens (spaces, lots of punctuation)
        if " " in value:
            return False
        alnum_ratio = sum(c.isalnum() or c in "-_.~+/=" for c in value) / len(value)
        return alnum_ratio >= 0.8

File: Other Versions/Go/scanners/vscode_extensions.go
package scanners

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"aihound/core"
	"aihound/remediation"
)

var vscodeExcludedExtensions = map[string]bool{
	"github.copilot":        true,
	"github.copilot-chat":   true,
	"saoudrizwan.claude-dev": true, // Cline
}

const vscodeMaxJSONSize = 1024 * 1024 // 1 MB
const vscodeMaxDirDepth = 3
const vscodeMaxJSONDepth = 8

var vscodeSecretKeySubstrings = []string{"token", "key", "secret", "password", "apikey", "auth"}

const vscodeRemediation = "Use VS Code's SecretStorage API or OS keychain for extension credentials"

type vscodeExtensionsScanner struct{}

func init() {
	Register(&vscodeExtensionsScanner{})
}

func (s *vscodeExtensionsScanner) Name() string        { return "VS Code Extensions" }
func (s *vscodeExtensionsScanner) Slug() string         { return "vscode-extensions" }
func (s *vscodeExtensionsScanner) IsApplicable() bool   { return true }

func (s *vscodeExtensionsScanner) Scan(showSecrets bool) core.ScanResult {
	result := core.ScanResult{ScannerName: s.Name(), Platform: core.DetectPlatform().String()}
	plat := core.DetectPlatform()

	for _, base := range s.getGlobalStorageRoots(plat) {
		s.scanGlobalStorage(base, &result, showSecrets)
	}

	return result
}

func (s *vscodeExtensionsScanner) getGlobalStorageRoots(plat core.Platform) []string {
	var roots []string

	if plat == core.PlatformLinux || plat == core.PlatformWSL {
		roots = append(roots, filepath.Join(core.GetXDGConfig(), "Code", "User", "globalStorage"))
	}
	if plat == core.PlatformMacOS {
		roots = append(roots, filepath.Join(core.GetHome(), "Library", "Application Support", "Code", "User", "globalStorage"))
	}
	if plat == core.PlatformWindows {
		if appdata := core.GetAppData(); appdata != "" {
			roots = append(roots, filepath.Join(appdata, "Code", "User", "globalStorage"))
		}
	}

	if plat == core.PlatformWSL {
		if appdata := core.GetAppData(); appdata != "" {
			roots = append(roots, filepath.Join(appdata, "Code", "User", "globalStorage"))
		}
		if winHome := core.GetWSLWindowsHome(); winHome != "" {
			roots = append(roots, filepath.Join(winHome, "AppData", "Roaming", "Code", "User", "globalStorage"))
		}
	}

	// De-duplicate
	seen := map[string]bool{}
	var unique []string
	for _, r := range roots {
		if !seen[r] {
			seen[r] = true
			unique = append(unique, r)
		}
	}
	return unique
}

func (s *vscodeExtensionsScanner) scanGlobalStorage(base string, result *core.ScanResult, showSecrets bool) {
	info, err := os.Stat(base)
	if err != nil || !info.IsDir() {
		return
	}

	entries, err := os.ReadDir(base)
	if err != nil {
		return
	}

	for _, entry := range entries {
		if !entry.IsDir() {
			continue
		}
		extID := entry.Name()
		if vscodeExcludedExtensions[extID] {
			continue
		}

		extDir := filepath.Join(base, extID)
		jsonFiles := s.iterJSONFiles(extDir, 0)
		for _, jf := range jsonFiles {
			s.scanExtensionJSON(jf, extID, result, showSecrets)
		}
	}
}

func (s *vscodeExtensionsScanner) iterJSONFiles(directory string, depth int) []string {
	if depth > vscodeMaxDirDepth {
		return nil
	}
	children, err := os.ReadDir(directory)
	if err != nil {
		return nil
	}

	var out []string
	for _, child := range children {
		childPath := filepath.Join(directory, child.Name())
		if child.IsDir() {
			out = append(out, s.iterJSONFiles(childPath, depth+1)...)
			continue
		}
		if strings.ToLower(filepath.Ext(child.Name())) != ".json" {
			continue
		}
		info, err := child.Info()
		if err != nil || info.Size() > vscodeMaxJSONSize {
			continue
		}
		out = append(out, childPath)
	}
	return out
}

func (s *vscodeExtensionsScanner) scanExtensionJSON(path, extensionID string, result *core.ScanResult, showSecrets bool) {
	raw, err := os.ReadFile(path)
	if err != nil {
		return
	}

	var data interface{}
	if err := json.Unmarshal(raw, &data); err != nil {
		return
	}

	perms := core.GetFilePermissions(path)
	owner := core.GetFileOwner(path)
	mtime := core.GetFileMtime(path)

	stalenessNote := ""
	if mt := core.GetFileMtimeTime(path); !mt.IsZero() {
		stalenessNote = "File last modified: " + core.DescribeStaleness(mt)
	}

	storage := core.PlaintextJSON

	walkOut := s.walk(data, "", 0)
	for _, entry := range walkOut {
		value, ok := entry.value.(string)
		if !ok {
			continue
		}

		leaf := entry.keyPath
		if idx := strings.LastIndex(entry.keyPath, "."); idx >= 0 {
			leaf = entry.keyPath[idx+1:]
		}
		leafLower := strings.ToLower(leaf)
		matched := false
		for _, sub := range vscodeSecretKeySubstrings {
			if strings.Contains(leafLower, sub) {
				matched = true
				break
			}
		}
		if !matched {
			continue
		}

		if !vscodeValueLooksSecret(value) {
			continue
		}

		notes := []string{
			fmt.Sprintf("Extension: %s", extensionID),
			fmt.Sprintf("JSON path: %s", entry.keyPath),
		}
		if stalenessNote != "" {
			notes = append(notes, stalenessNote)
		}

		rawValue := ""
		if showSecrets {
			rawValue = value
		}

		result.Findings = append(result.Findings, core.CredentialFinding{
			ToolName:        s.Name(),
			CredentialType:  fmt.Sprintf("%s:%s", extensionID, entry.keyPath),
			StorageType:     storage,
			Location:        path,
			Exists:          true,
			RiskLevel:       core.AssessRisk(storage, path),
			ValuePreview:    core.MaskValue(value, showSecrets),
			RawValue:        rawValue,
			FilePermissions: perms,
			FileOwner:       owner,
			FileModified:    mtime,
			Remediation:     vscodeRemediation,
			RemediationHint: remediation.HintManual(vscodeRemediation),
			Notes:           notes,
		})
	}
}

type vscodeWalkEntry struct {
	keyPath string
	value   interface{}
}

func (s *vscodeExtensionsScanner) walk(data interface{}, prefix string, depth int) []vscodeWalkEntry {
	if depth > vscodeMaxJSONDepth {
		return nil
	}
	var out []vscodeWalkEntry

	switch v := data.(type) {
	case map[string]interface{}:
		for k, val := range v {
			newPrefix := k
			if prefix != "" {
				newPrefix = prefix + "." + k
			}
			switch val.(type) {
			case map[string]interface{}, []interface{}:
				out = append(out, s.walk(val, newPrefix, depth+1)...)
			default:
				out = append(out, vscodeWalkEntry{keyPath: newPrefix, value: val})
			}
		}
	case []interface{}:
		for i, val := range v {
			newPrefix := fmt.Sprintf("[%d]", i)
			if prefix != "" {
				newPrefix = fmt.Sprintf("%s[%d]", prefix, i)
			}
			switch val.(type) {
			case map[string]interface{}, []interface{}:
				out = append(out, s.walk(val, newPrefix, depth+1)...)
			default:
				out = append(out, vscodeWalkEntry{keyPath: newPrefix, value: val})
			}
		}
	}
	return out
}

func vscodeValueLooksSecret(value string) bool {
	if len(value) < 20 {
		return false
	}
	if strings.HasPrefix(value, "/") || strings.HasPrefix(value, "\\") ||
		strings.HasPrefix(value, "http://") || strings.HasPrefix(value, "https://") ||
		strings.HasPrefix(value, "file://") {
		return false
	}
	if strings.Contains(value, " ") {
		return false
	}
	// alnum + -_.~+/=
	allowed := 0
	for _, c := range value {
		if (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || (c >= '0' && c <= '9') ||
			c == '-' || c == '_' || c == '.' || c == '~' || c == '+' || c == '/' || c == '=' {
			allowed++
		}
	}
	ratio := float64(allowed) / float64(len(value))
	return ratio >= 0.8
}
